### PR TITLE
WIP: Infrastructure for fetchable RAND.

### PR DESCRIPTION
**Pull request by paulidale**
This adds the infrastructure for a RAND algorithm type.  These will include the DRBGs and seeding sources.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: apps/list.c [hunk 1]
```diff
@@ -244,6 +244,56 @@ static void list_kdfs(void)
...
+static int rand_cmp(const EVP_RAND * const *a, const EVP_RAND * const *b)
+{
+    int ret = strcasecmp(EVP_RAND_name(*a), EVP_RAND_name(*b));
+
+    if (ret == 0)
+        ret = strcmp(OSSL_PROVIDER_name(EVP_RAND_provider(*a)),
+                     OSSL_PROVIDER_name(EVP_RAND_provider(*b)));
+
+    return ret;
+}
```
Comment (levitte) [inline apps/list.c]: I've some work going on in #9979 that touches on this...  I'd like to get that through first if possible

Comment (paulidale) [reply apps/list.c]: Sure, no rush here.  There is plenty of other things to be done on the RAND front.

I could remove the `list.c` changes to avoid interfering.

File: doc/man7/EVP_RAND-TEST_RAND.pod [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+=head1 DESCRIPTION
+
+A test random number generator which should never be used for production.
+It accepts seed material and regurgitates this verbatim instead of generating
+anything.
+
+=head2 Identity
+
+"TEST-RAND" is the name for this implementation; it
+can be used with the EVP_RANDF_fetch() function.
```
Comment (levitte) [inline doc/man7/EVP_RAND-TEST_RAND.pod]: RANDF?

File: include/openssl/evperr.h [hunk 3]
```diff
@@ -10,12 +10,6 @@
 
 #ifndef OPENSSL_EVPERR_H
 # define OPENSSL_EVPERR_H
-# pragma once
-
-# include <openssl/macros.h>
-# if !OPENSSL_API_3
-#  define HEADER_EVPERR_H
-# endif
```
Comment (levitte) [inline include/openssl/evperr.h]: Why did you remove this?

File: include/openssl/core_names.h [hunk 4]
```diff
@@ -143,6 +143,21 @@ extern "C" {
...
+/* RAND parameters */
+#define OSSL_RAND_PARAM_STRENGTH    "strength"                /* unsigned int */
+#define OSSL_RAND_PARAM_SEED        "seed"                    /* octet string */
+#define OSSL_RAND_PARAM_PERSONALISE "personalise"             /* octet string */
+#define OSSL_RAND_PARAM_CIPHER      OSSL_ALG_PARAM_CIPHER     /* utf8 string */
+#define OSSL_RAND_PARAM_DIGEST      OSSL_ALG_PARAM_DIGEST     /* utf8 string */
+#define OSSL_RAND_PARAM_PROPERTIES  OSSL_ALG_PARAM_PROPERTIES /* utf8 string */
+
+/* Known RAND names */
+#define OSSL_RAND_NAME_SEED         "seed"
```
Comment (levitte) [inline include/openssl/core_names.h]: Is this for seed source methods?

Comment (levitte) [reply include/openssl/core_names.h]: Will there be only one?

File: apps/list.c (hunk 1 above)
Comment (levitte) [reply apps/list.c]: I have today's work cut out, then 😉

File: include/openssl/evperr.h (hunk 3 above)
Comment (paulidale) [reply include/openssl/evperr.h]: This happened during `make update`, I didn't manually touch this file.

File: include/openssl/core_names.h (hunk 4 above)
Comment (paulidale) [reply include/openssl/core_names.h]: Yes, this will be for _our_ seeding from the whatever was configured.  I'm happy to rename it, that part isn't done yet & I'm not going finish before I fly out for the F2F.

File: doc/man7/EVP_RAND-TEST_RAND.pod (hunk 2 above)
Comment (paulidale) [reply doc/man7/EVP_RAND-TEST_RAND.pod]: Fixed.

File: include/openssl/evperr.h (hunk 3 above)
Comment (paulidale) [reply include/openssl/evperr.h]: Put back.

File: apps/list.c (hunk 1 above)
Comment (paulidale) [reply apps/list.c]: I can drop this out of the PR if it helps.

**Comment by paulidale** [comment]
>It's interesting to see how EVP stuff kinda leaks into headers that aren't historically EVP, or new EVP headers forming. kdf.h, now rand.h. I don't mind, at all, 'cause evp.h is a bit over inflated. Maybe we should make that a more consistent thing, though...

I'd noticed this as well, refreshing _evp.h_ is something I'd like to do.

I didn't want to increase the entropy that is _evp.h_ and creating _evp_rand.h_ seemed kind of awkward.  Especially in light of _kdf.h_ not being _evp_kdf.h_.  Would an `evp` directory make sense with the top level _evp.h_ including everything?  Any fix would be a different PR I suspect.

File: include/openssl/evperr.h (hunk 3 above)
Comment (levitte) [reply include/openssl/evperr.h]: Hmmm...  @mspncp, you might want to have a look at what happens with a make update.

**Comment by levitte** [comment]
> Would an `evp` directory make sense with the top level _evp.h_ including everything?

Please, let's avoid the directory tree idea.  I think that _kdf.h_ paves the way nicely, and we could have _evp.h_ simply include whatever we see fit (for example _kdf.h_).

> Any fix would be a different PR I suspect.

Yes.

File: include/openssl/core_names.h (hunk 4 above)
Comment (paulidale) [reply include/openssl/core_names.h]: The reason seed sources are included here is because of the way seeding DRBGs works.  A DRBG has a parent DRBG pointer.  If this pointer is set, the parent is used as a seed source.  If the pointer is NULL, the built in source is used instead.  This inhibits having more than one seed source.  EVP_RAND is adding a layer that will allow either to be used for seeding.

**Comment by paulidale** [comment]
Okay, I'll leave the new declarations in rand.h.

File: include/openssl/core_names.h (hunk 4 above)
Comment (levitte) [reply include/openssl/core_names.h]: `ossl-seed`?

Comment (paulidale) [reply include/openssl/core_names.h]: I'm fine with that.

Comment (paulidale) [reply include/openssl/core_names.h]: and done.

File: providers/implementations/rands/test_rand.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    size_t current;
+    unsigned char *buf;
+} TEST_RAND;
+
+static void *test_rand_new(void *provctx)
+{
+    TEST_RAND *ctx;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL)
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
```
Comment (levitte) [inline providers/implementations/rands/test_rand.c]: You need to include `<openssl/err.h>`

Comment (paulidale) [reply providers/implementations/rands/test_rand.c]: Added.

**Comment by levitte** [comment]
Travis failure relevant

File: include/crypto/evp.h [hunk 6]
```diff
@@ -178,6 +178,30 @@ extern const EVP_KDF ss_kdf_meth;
...
+    OSSL_OP_rand_get_params_fn *get_params;
+    OSSL_OP_rand_get_ctx_params_fn *get_ctx_params;
+    OSSL_OP_rand_set_ctx_params_fn *set_ctx_params;
+};
+
+extern const EVP_RAND seed_rand_meth;
+extern const EVP_RAND ctr_drbg_rand_meth;
+extern const EVP_RAND digest_drbg_rand_meth;
+extern const EVP_RAND hmac_drbg_rand_meth;
+extern const EVP_RAND test_rand_meth;
```
Comment (levitte) [inline include/crypto/evp.h]: These aren't needed, are they?

File: include/openssl/core_numbers.h [hunk 7]
```diff
@@ -312,6 +313,33 @@ OSSL_CORE_MAKE_FUNC(int, OP_kdf_get_ctx_params,
...
+OSSL_CORE_MAKE_FUNC(int, OP_rand_reseed, (void *rctx))
+OSSL_CORE_MAKE_FUNC(const OSSL_PARAM *, OP_rand_gettable_params, (void))
+OSSL_CORE_MAKE_FUNC(const OSSL_PARAM *, OP_rand_gettable_ctx_params, (void))
+OSSL_CORE_MAKE_FUNC(const OSSL_PARAM *, OP_rand_settable_ctx_params, (void))
+OSSL_CORE_MAKE_FUNC(int, OP_rand_get_params, (OSSL_PARAM params[]))
+OSSL_CORE_MAKE_FUNC(int, OP_rand_get_ctx_params,
+                    (void *rctx, OSSL_PARAM params[]))
+OSSL_CORE_MAKE_FUNC(int, OP_rand_set_ctx_params,
+                    (void *rctx, const OSSL_PARAM params[]))
+
```
Comment (levitte) [inline include/openssl/core_numbers.h]: `doc/man7/provider-rand.h`?

File: include/crypto/evp.h (hunk 6 above)
Comment (paulidale) [reply include/crypto/evp.h]: Not yet, but they will be once the rest of the methods are added.
I'm fine deleting them for the moment.

File: include/openssl/core_numbers.h (hunk 7 above)
Comment (paulidale) [reply include/openssl/core_numbers.h]: Ah, missed this one.  The new stuff has [tentacles](http://hplhs.org/freebies/audiosamples/tentaclessample.mp3) everywhere....

File: include/crypto/evp.h (hunk 6 above)
Comment (levitte) [reply include/crypto/evp.h]: I'm lost. Why are they needed?

File: include/openssl/core_numbers.h (hunk 7 above)
Comment (paulidale) [reply include/openssl/core_numbers.h]: Interestingly, there is no provider-kdf documentation file :(

Comment (paulidale) [reply include/openssl/core_numbers.h]: Done, crudely.

File: include/crypto/evp.h (hunk 6 above)
Comment (paulidale) [reply include/crypto/evp.h]: They are not needed at the moment.
They will be needed once the DRBGs and seed source are moved across to this framework.

`test_rand_meth` is needed already.


I'm quite happy to delete this and put them back when the rest is done (after the F2F most likely).

**Comment by paulidale** [comment]
Travis is happy now :)

File: include/crypto/evp.h (hunk 6 above)
Comment (levitte) [reply include/crypto/evp.h]: You don't seem to understand. With fetched algos, *there should be no needed for instances like this*, they are inherently legacy.

Comment (paulidale) [reply include/crypto/evp.h]: This is an internal header file which provides access for the providers to declare the implementations.  Where else should they go?

Comment (levitte) [reply include/crypto/evp.h]: That sounds like you're mixing this up with `providers/implementations/include/prov/implementations.h`.

BTW, I just noticed that the legacy KDF stuff hasn't been cleaned away, so that provides an example of what I'm saying:

```
: ; git grep tls1_prf_kdf_meth
include/crypto/evp.h:extern const EVP_KDF tls1_prf_kdf_meth;
: ; 
```

No definition, or use, of `tls1_prf_kdf_meth`

Comment (paulidale) [reply include/crypto/evp.h]: I understand now.  The KDF stuff is best left to another PR.  I can fix the rand stuff.

Comment (paulidale) [reply include/crypto/evp.h]: I've removed the errant definitions -- they weren't used.

I've removed the KDFs as a separate commit.

Comment (levitte) [reply include/crypto/evp.h]: [g] #10170

Comment (paulidale) [reply include/crypto/evp.h]: I removed the commit from here and have approved your commit.

**Comment by paulidale** [comment]
Rebased after #10170

File: crypto/evp/rand.c [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+{
+    if (outlen == 0)
+        return 1;
+    if (ctx == NULL || out == NULL)
+        return 0;
+    return ctx->meth->get_bytes(ctx->data, out, outlen);
+}
+
+/*
+ * The {get,set}_params functions return 1 if there is no corresponding
```
Comment (slontis) [inline crypto/evp/rand.c]: maybe they should return 0, (or -1) for not supported?

File: crypto/evp/rand.c [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+            if (rand->set_ctx_params != NULL)
+                break;
+            rand->set_ctx_params = OSSL_get_OP_rand_set_ctx_params(fns);
+            break;
+        }
+    }
+    if (fnrandcnt != 1 || fnctxcnt != 2) {
+        /*
+         * In order to be a consistent set of functions we must have at least
+         * a derive function, and a complete set of context management
```
Comment (slontis) [inline crypto/evp/rand.c]: probably update this block copied comment :)

File: doc/man3/EVP_RAND.pod [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+ int EVP_RAND_CTX_get_params(EVP_RAND_CTX *ctx, OSSL_PARAM params[]);
+ int EVP_RAND_CTX_set_params(EVP_RAND_CTX *ctx, const OSSL_PARAM params[]);
+ const OSSL_PARAM *EVP_RAND_gettable_params(const EVP_RAND *rand);
+ const OSSL_PARAM *EVP_RAND_gettable_ctx_params(const EVP_RAND *rand);
+ const OSSL_PARAM *EVP_RAND_settable_ctx_params(const EVP_RAND *rand);
+ const OSSL_PROVIDER *EVP_RAND_provider(const EVP_RAND *rand);
+
+=head1 DESCRIPTION
+
+The EVP RAND routines are a high level interface to Key Derivation Function
```
Comment (slontis) [inline doc/man3/EVP_RAND.pod]: KDF :)

File: doc/man3/EVP_RAND.pod [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+=item "maclen" (B<OSSL_RAND_PARAM_MAC_SIZE>) <unsigned integer>
+
+Used by implementations that use a MAC with a variable output size (KMAC).
+For those RAND implementations that support it, this parameter
+sets the MAC output size.
+
+The default value, if any, is implementation dependent.
+The length must never exceed what can be given with a B<size_t>.
+
+=item "maxmem_bytes" (B<OSSL_RAND_PARAM_SCRYPT_MAXMEM>) <unsigned integer>
```
Comment (slontis) [inline doc/man3/EVP_RAND.pod]: What??

File: doc/man3/EVP_RAND.pod [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+NULL on error.
+
+EVP_MAC_up_ref() returns 1 on success, 0 on error.
+
+EVP_RAND_CTX_new() returns either the newly allocated
+C<EVP_RAND_CTX> structure or C<NULL> if an error occurred.
+
+EVP_RAND_CTX_free() and EVP_RAND_reset() do not return a value.
+
+EVP_RAND_strenth() returns the output strength.
```
Comment (slontis) [inline doc/man3/EVP_RAND.pod]: strength

File: doc/man7/EVP_RAND-TEST_RAND.pod [hunk 13]
```diff
@@ -0,0 +1,88 @@
+=pod
+
+=head1 NAME
+
+EVP_RAND-TEST_RAND - The test EVP_RAND implementation
+
+=head1 DESCRIPTION
+
+A test random number generator which should never be used for production.
```
Comment (slontis) [inline doc/man7/EVP_RAND-TEST_RAND.pod]: hmm the term production might be misleading - we may use this internally in self tests

File: doc/man7/EVP_RAND-TEST_RAND.pod [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+=head2 Supported parameters
+
+The supported parameters are:
+
+=over 4
+
+=item "strength" (B<OSSL_RAND_PARAM_STRENGTH>) <unsigned integer>
+
+Get the security strength of the test random number generator.
+This parameter cannot be set and will always have the value zero.
```
Comment (slontis) [inline doc/man7/EVP_RAND-TEST_RAND.pod]: Not sure the 2nd statement is needed.

File: doc/man7/provider-rand.pod [hunk 15]
```diff
@@ -0,0 +1,193 @@
+=pod
+
+=head1 NAME
+
+provider-rand - The random numbe generator library E<lt>-E<gt> provider
```
Comment (slontis) [inline doc/man7/provider-rand.pod]: number

File: doc/man7/provider-rand.pod [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+  * pointers in OSSL_DISPATCH arrays.
+  */
+
+ /* Context management */
+ void *OP_rand_newctx(void *provctx);
+ void OP_rand_freectx(void *rctx);
+
+ /* Encryption/decryption */
+ int OP_rand_reseed(void *rctx);
+ int OP_rand_get_bytes(void *rctx, unsigned char *out, size_t *outlen;
```
Comment (slontis) [inline doc/man7/provider-rand.pod]: );

File: include/openssl/evperr.h (hunk 3 above)
Comment (mspncp) [reply include/openssl/evperr.h]: Sorry for the delay @levitte, I only noticed your message today. Too much input currently. I will take a look as soon as possible.

File: crypto/evp/rand.c (hunk 8 above)
Comment (paulidale) [reply crypto/evp/rand.c]: This is the same as is in `mac_lib.c` and `kdf_lib.c`.  It might not be ideal but it is consistent.

We need to canonically define what happens when some of the params are not known.

File: crypto/evp/rand.c (hunk 9 above)
Comment (paulidale) [reply crypto/evp/rand.c]: Done

File: doc/man3/EVP_RAND.pod (hunk 10 above)
Comment (paulidale) [reply doc/man3/EVP_RAND.pod]: Yeah, fixed.

File: doc/man3/EVP_RAND.pod (hunk 11 above)
Comment (paulidale) [reply doc/man3/EVP_RAND.pod]: This whole man page needs work....

File: doc/man3/EVP_RAND.pod (hunk 12 above)
Comment (paulidale) [reply doc/man3/EVP_RAND.pod]: Fixed.

File: doc/man7/EVP_RAND-TEST_RAND.pod (hunk 13 above)
Comment (paulidale) [reply doc/man7/EVP_RAND-TEST_RAND.pod]: Decent point.  Reworded.

File: doc/man7/EVP_RAND-TEST_RAND.pod (hunk 14 above)
Comment (paulidale) [reply doc/man7/EVP_RAND-TEST_RAND.pod]: I'll leave it as "this param cannot be set".

File: doc/man7/provider-rand.pod (hunk 15 above)
Comment (paulidale) [reply doc/man7/provider-rand.pod]: Fixed.

File: doc/man7/provider-rand.pod (hunk 16 above)
Comment (paulidale) [reply doc/man7/provider-rand.pod]: An unhappy winking face????

Fixed.

File: util/libcrypto.num [hunk 17]
```diff
@@ -4857,3 +4857,21 @@ X509_LOOKUP_store                       4973	3_0_0	EXIST::FUNCTION:
...
+EVP_RAND_get_params                     4985	3_0_0	EXIST::FUNCTION:
+EVP_RAND_CTX_get_params                 4986	3_0_0	EXIST::FUNCTION:
+EVP_RAND_CTX_set_params                 4987	3_0_0	EXIST::FUNCTION:
+EVP_RAND_fetch                          4988	3_0_0	EXIST::FUNCTION:
+EVP_RAND_up_ref                         4989	3_0_0	EXIST::FUNCTION:
+EVP_RAND_free                           4990	3_0_0	EXIST::FUNCTION:
+EVP_RAND_gettable_params                4991	3_0_0	EXIST::FUNCTION:
+EVP_RAND_gettable_ctx_params            4992	3_0_0	EXIST::FUNCTION:
+EVP_RAND_settable_ctx_params            4993	3_0_0	EXIST::FUNCTION:
+EVP_RAND_do_all_ex                      4994	3_0_0	EXIST::FUNCTION:
```
Comment (levitte) [inline util/libcrypto.num]: Do another rebase and change those numbers to question marks

Comment (paulidale) [reply util/libcrypto.num]: It's planned.  I need to rethink how I'm doing some of this.  The three DRBGs share a lot of common code to do with maximum sizes and reseed.  None of this is appropriate to seed sources.

File: providers/common/include/prov/rand_local.h [hunk 18]
```diff
@@ -0,0 +1,360 @@
+/*
```
Comment (levitte) [inline providers/common/include/prov/rand_local.h]: Why do you name this file "rand_*local*.h"?

Comment (paulidale) [reply providers/common/include/prov/rand_local.h]: Because everything is mixed up in my head still isn't a good answer :)
I'll rename it.

Comment (paulidale) [reply providers/common/include/prov/rand_local.h]: It's old residue too.  The equivalent is in implementations/rand.

File: include/openssl/core_numbers.h [hunk 19]
```diff
@@ -313,6 +324,65 @@ OSSL_CORE_MAKE_FUNC(int, OP_kdf_get_ctx_params,
...
+/*
+ * Typedef to allow random number generators to be chained between providers
+ * without libcrypto being in the middle.  This requires the ability for one
+ * provider to directly access callbacks from another.
+ */
```
Comment (levitte) [inline include/openssl/core_numbers.h]: I don't see this reflected in the code. All actual calls that I can find happen in libcrypto. What am I missing?
If all the calls are *supposed* to happen in libcrypto, then I don't see why this structure is defined here...
(mind you, a structure like this being shared between libcrypto and providers is a big no-no to begin with)

Comment (paulidale) [reply include/openssl/core_numbers.h]: The connection happens in libcrypto, the calling will be in the provider.  Haven't got that far yet...

The existence of this structure boils down to the question of being able to fetch OSSL_DISTPACH tables or not.

Comment (paulidale) [reply include/openssl/core_numbers.h]: The `get_entropy` call in the DRBG will need to call the DRBG's parent for seed material or something similar.

Comment (levitte) [reply include/openssl/core_numbers.h]: So basically, what's currently happening is that you pass a method structure between libcrypto and provider.  This violates the principle we've had so far, not to share exactly this type of stuff between libcrypto and provider; any change in this structure, or any of the signatures, and that will break compatibility.  We have enough and more flexible shared knowledge through `OSSL_ALGORITHM` / `OSSL_DISPATCH`, do pass that instead and let the provider build up its own method structure.

Comment (paulidale) [reply include/openssl/core_numbers.h]: Agreed.

I'm not passing all the methods across and some in this structure won't be required (I added them in the wrong place).

File: include/openssl/evperr.h (hunk 3 above)
Comment (mspncp) [reply include/openssl/evperr.h:18]: > Hmmm... @mspncp, you might want to have a look at what happens with a make update.

@levitte the fix is straighforward and for you it's probably a five minute task: The mkerr.pl script needs to be adjusted to add the legacy guard, but not unconditionally; only if the file already has a legacy guard. I've written down the details in #10569. With my limited perl skills , it would take me significantly longer to figure it out.

**Comment by levitte** [comment]
I assume this was abandoned in favour of #11682

**Comment by mspncp** [comment]
Yup.

**Comment by paulidale** [comment]
Yes, it has been abandoned.  The two are done along similar lines, but I did the later PR in manageable sized steps knowing more accurately what the final solution needed to look like.  This PR was developed in a far less controlled manner and I worry I've dropped something important during the process.
